crypto/fipsmodule/service_indicator/service_indicator_test.cc

```diff
@@ -63,6 +67,242 @@ static const uint8_t kAESCBCCiphertext[64] = {
+  unsigned digest_len;
+
+  // Test running the EVP_Digest interfaces one by one directly, and check
+  // |EVP_DigestFinal_ex| for approval at the end.
+  ASSERT_TRUE(EVP_DigestInit_ex(ctx.get(), digestTestVector.md.func(), nullptr));
+  ASSERT_TRUE(EVP_DigestUpdate(ctx.get(), digestTestVector.input, sizeof(digestTestVector.input)));
+  CALL_SERVICE_AND_CHECK_APPROVED(approved, ASSERT_TRUE(EVP_DigestFinal_ex(ctx.get(), digest.get(), &digest_len)));
+  ASSERT_EQ(approved, digestTestVector.expect_approved);
+  if(approved == AWSLC_APPROVED) {
+    ASSERT_TRUE(check_test(digestTestVector.expected_digest, digest.get(), digest_len, digestTestVector.md.name));
+  }
+
+  // Test using the one-shot |EVP_Digest| function for approval.
+  CALL_SERVICE_AND_CHECK_APPROVED(approved, EVP_Digest(digestTestVector.input, sizeof(digestTestVector.input),
+                                               digest.get(), &digest_len, digestTestVector.md.func(), nullptr));
+  ASSERT_EQ(approved, digestTestVector.expect_approved);
+  if(approved == AWSLC_APPROVED) {
+    ASSERT_TRUE(check_test(digestTestVector.expected_digest, digest.get(), digest_len, digestTestVector.md.name));
+  }
+
+  // Test using the one-shot API for approval.
+  CALL_SERVICE_AND_CHECK_APPROVED(approved, digestTestVector.md.one_shot_func(digestTestVector.input, sizeof(digestTestVector.input), digest.get()));
+  ASSERT_EQ(approved, digestTestVector.expect_approved);
+  if(approved == AWSLC_APPROVED) {
+    ASSERT_TRUE(check_test(digestTestVector.expected_digest, digest.get(), digestTestVector.md.length, digestTestVector.md.name));
+  }
+}
+
+struct HMACTestVector {
+  // func is the hash function for HMAC to test.
+  const EVP_MD *(*func)(void);
+  // input is a NUL-terminated string to hash.
+  const uint8_t *input;
+  // expected_digest is the expected digest.
+  const uint8_t *expected_digest;
+  // expected to be approved or not.
+  const int expect_approved;
+} kHMACTestVectors[] = {
+    { EVP_sha1, kPlaintext, kHMACOutput_sha1, AWSLC_APPROVED },
+    { EVP_sha224, kPlaintext, kHMACOutput_sha224, AWSLC_APPROVED },
+    { EVP_sha256, kPlaintext, kHMACOutput_sha256, AWSLC_APPROVED },
+    { EVP_sha384, kPlaintext, kHMACOutput_sha384, AWSLC_APPROVED },
+    { EVP_sha512, kPlaintext, kHMACOutput_sha512, AWSLC_APPROVED },
+    { EVP_sha512_256, kPlaintext, nullptr, AWSLC_NOT_APPROVED }
+};
+
+class HMAC_ServiceIndicatorTest : public testing::TestWithParam<HMACTestVector> {};
+
+INSTANTIATE_TEST_SUITE_P(All, HMAC_ServiceIndicatorTest, testing::ValuesIn(kHMACTestVectors));
+
+TEST_P(HMAC_ServiceIndicatorTest, HMACTest) {
+  const HMACTestVector &hmacTestVector = GetParam();
+
+  int approved = AWSLC_NOT_APPROVED;
+  const uint8_t kHMACKey[64] = {0};
+  std::string digest_str;
+  const EVP_MD *digest = hmacTestVector.func();
+  std::vector<uint8_t> key(kHMACKey, kHMACKey + sizeof(kHMACKey));
+  std::vector<uint8_t> input(hmacTestVector.input, hmacTestVector.input + sizeof(hmacTestVector.input));
```

Comment:
Why did you need this `input` vector? Also, because I don't know, why `key` and `input` are vectors while not `mac` (nor `digest` in the test above)?

Reply:
It was from the original test I copied from. Changed all of them to vectors.